py/dml/c_backend.py

```diff
@@ -1093,14 +1097,145 @@ def generate_simple_events(device):
+            out('%s = _args;\n' % (args_decl,))
+        indices_lit = 'indices' if info.dimensions else None
+        args_lit = 'args' if info.args_type else None
+        msg_lit = 'msg' if info.param_to_msg_comp else None
+        info.generate_callback_call(indices_lit, args_lit, msg_lit)
+        out('}\n\n', preindent = -1)
+        splitting_point()
+
+        if info.has_serialized_args:
+            start_function_definition(
+                f'attr_value_t {info.cident_args_serializer}('
+                + 'const void *_args)')
+            out('{\n', postindent = 1)
+            if info.args_type:
+                args_type_ptr = TPtr(conv_const(True, info.args_type))
+                out(f'{args_type_ptr.declaration("args")} = _args;\n')
+                args_expr = mkDereference(site, mkLit(site, 'args', args_type_ptr))
+            else:
+                args_expr = None
+            out('attr_value_t out;\n')
+            out_expr = mkLocalVariable(
+                site, LocalSymbol('out', 'out', attr_value_t, site=site))
+            info.generate_args_serializer(site, args_expr, out_expr)
+            out('return out;\n')
+            out('}\n\n', preindent = -1)
+            splitting_point()
+
+            start_function_definition(
+                f'set_error_t {info.cident_args_deserializer}('
+                + 'attr_value_t val, void *_out)')
+            out('{\n', postindent = 1)
+            out('set_error_t _success UNUSED = Sim_Set_Ok;\n')
+
+            if info.args_type:
+                out(f'{TPtr(info.args_type).declaration("out")} = _out;\n')
+                out_expr = mkDereference(site, mkLit(site, 'out',
+                                                     TPtr(info.args_type)))
+            else:
+                out_expr = None
+            def error_out(exc, msg):
+                stmts = []
+                stmts.append(mkInline(site, f'_success = {exc};'))
+                if msg is not None:
+                    stmts.append(
+                        mkInline(site, f'SIM_attribute_error("{msg}");'))
+                stmts.append(mkInline(site, 'goto _exit;'))
+                return stmts
+            val_expr = mkLit(site, 'val', attr_value_t)
+
+            info.generate_args_deserializer(site, val_expr, out_expr,
+                                            error_out)
+
+            out('_exit:\n')
+            out('return _success;\n')
+            out('}\n\n', preindent = -1)
+            splitting_point()
+
+    reset_line_directive()
+
+    if dml.globals.after_on_hook_infos:
+        items = (
+            '{%s, %s, %s, %s, %s, %d}'
+            % ((string_literal(info.string_key), info.cident_callback)
+               + ((info.cident_args_serializer, info.cident_args_deserializer)
+                  if info.has_serialized_args else ('NULL', 'NULL'))
+               + (f'sizeof({info.args_type.declaration("")})'
+                  if info.args_type else '0', info.parent.uniq))
```

Comment:
this looks pretty odd, in particular tuple arithmetic, but I can see why you do it (moving out conditionals to separate statements could improve formatting, but at the cost of abandoning generators). Did you consider expressing it as a sum of three f-strings would look better? (would probably be an improvement code-wise, but at the cost of disguising that the result is a 6-element initializer)

Reply:
The `%` syntax was actually for readability and my sanity. When you have a bunch of concatenated `f-strings`, I feel their readability easily becomes _horrible_, especially when the embedded expressions get long. It makes it too easy to lose track of what part of the string you're writing, too easy to mess up details about the string outside of the embedded expressions (for example, makes it too easy to forget writing a comma), and can make it extremely difficult to gauge what kind of string you're actually _formatting._ My stance on `%` vs f-strings has changed because of this: f-strings are superior for strings with a few number of short embedded expressions, but `%` is superior for strings with a large number of large embedded expressions.